code/datums/job_selection/job_candidate.dm

```diff
@@ -0,0 +1,195 @@
+RESTRICT_TYPE(/datum/job_candidate)
+
+/// A datum representing all the information needed to perform roundstart job
+/// selection, without actually needing clients. This allows for the creation
+/// of fake candidates, allowing testing of the role selection system independent
+/// of the number of actual players.
+/datum/job_candidate
+	VAR_PRIVATE/datum/character_save/active_character
+	VAR_PRIVATE/datum/job_ban_holder/jbh
+	VAR_PRIVATE/list/be_special = list()
+	VAR_PRIVATE/list/restricted_roles = list()
+
+	VAR_PRIVATE/account_age_in_days
+	VAR_PRIVATE/admin_rights = 0
+	VAR_PRIVATE/assigned_job_title = null
+	VAR_PRIVATE/exp
+	VAR_PRIVATE/is_admin = FALSE
+	VAR_PRIVATE/is_guest_key = FALSE
+	VAR_PRIVATE/latejoin = FALSE
+	VAR_PRIVATE/real_ckey
+	VAR_PRIVATE/return_to_lobby = FALSE
+	VAR_PRIVATE/special_role
+
+/datum/job_candidate/New()
+	active_character = new()
+	jbh = new()
+
+/datum/job_candidate/proc/to_string()
+	return real_ckey ? real_ckey : "(simulated) [active_character.real_name]"
+
+/datum/job_candidate/proc/is_assigned()
+	return assigned_job_title != null
+
+/datum/job_candidate/proc/assign_title(title)
+	assigned_job_title = title
+
+/datum/job_candidate/proc/return_to_lobby()
+	return_to_lobby = TRUE
+
+/datum/job_candidate/proc/restricted_from(datum/job/job)
+	return job.title in restricted_roles
+
+/datum/job_candidate/proc/matches_job_priority(datum/job/job, level)
+	return (active_character.GetJobDepartment(job, level) & job.flag)
+
+/datum/job_candidate/proc/has_special_role()
+	return special_role
+
+/datum/job_candidate/proc/alternate_spawn_option()
+	return active_character.alternate_option
+
+/datum/job_candidate/proc/has_restricted_roles()
+	return length(restricted_roles)
+
+/datum/job_candidate/proc/is_jobbanned(datum/job/job)
+	if(GLOB.configuration.jobs.guest_job_ban && is_guest_key)
+		return "Guest Job-ban"
+
+	if(job.title in jbh.job_bans)
+		var/datum/job_ban/ban = jbh.job_bans[job.title]
+		return ban.reason
+
+/datum/job_candidate/proc/is_incompatible_role(datum/job/job)
+	return job.title in restricted_roles
+
+/datum/job_candidate/proc/wants_job(datum/job/job, level)
+	return active_character.GetJobDepartment(job, level) & job.flag
+
+/datum/job_candidate/proc/is_account_old_enough(datum/job/job)
+	if(GLOB.configuration.jobs.restrict_jobs_on_account_age && isnum(account_age_in_days) && isnum(job.minimal_player_age))
+		return max(0, job.minimal_player_age - account_age_in_days) == 0
+
+	return TRUE
+
+/datum/job_candidate/proc/has_special(flag)
+	return flag in be_special
+
+/datum/job_candidate/proc/load_from_player(mob/new_player/player)
+	active_character = player.client.prefs.active_character
```

Comment:
Is it possible the client might be null here?

Reply:
i'll check for that just in case.